src/main/kotlin/org/wfanet/measurement/reporting/service/api/v2alpha/SetExpressionCompiler.kt

```diff
@@ -0,0 +1,434 @@
+ * bit positions of set bits constitute the primitive region. Ex: Given a Venn Diagram of 3
+ * reporting sets (rs0, rs1, rs2), a primitive region with an integer value equal to 3 has the bit
+ * representation b’011’. This means this primitive region is only covered by the intersection of
+ * rs0 and rs1 and not covered by rs2 (the order of the bit positions is from right to left). In
+ * other words, rs0 INTERSECT rs1 INTERSECT COMPLEMENT(rs2). Note that primitive regions are
+ * disjoint.
+ */
+private typealias PrimitiveRegion = ULong
+
+/**
+ * A union set is the union of a set of reporting sets, and it is represented by a bit
+ * representation of an integer. Only the reporting sets with IDs equal to the bit positions of set
+ * bits constitute the union set. Given a Venn Diagram of 3 reporting sets (rs0, rs1, rs2), a
+ * union-set with an integer value equal to 3 has the bit representation b’011’. This means this
+ * union-set is only covered by the union of rs0 and rs1 (the order of the bit positions is from
+ * right to left).
+ */
+private typealias UnionSet = ULong
+
+private typealias NumberReportingSets = Int
+
+/** A mapping from a [UnionSet] to its coefficient in the Venn diagram region decomposition. */
+private typealias UnionSetCoefficientMap = Map<UnionSet, Int>
+
+/**
+ * A mapping for cardinality computation from a [PrimitiveRegion] to its decomposition in terms of
+ * union-sets represented by [UnionSetCoefficientMap]. Take a case of 3 reporting sets (rs0, rs1,
+ * rs2) as an example. A primitive region with its value equal to 3 (b'011') means rs0 INTERSECT rs1
+ * INTERSECT COMPLEMENT(rs2). The decomposition of the cardinality of the region =
+ * PrimitiveRegionToUnionSetCoefficientMap\[region\] = {4: -1, 5: 1, 6: 1, 7: -1}, i.e. |union-set5|
+ * + |union-set6| - |union-set4| - |union-set7|.
+ */
+private typealias PrimitiveRegionToUnionSetCoefficientMap =
+  MutableMap<PrimitiveRegion, UnionSetCoefficientMap>
+
+/**
+ * A memory cache that stores the Venn diagram region cardinality decompositions for different
+ * numbers of reporting sets.
+ */
+private typealias PrimitiveRegionCache =
+  MutableMap<NumberReportingSets, PrimitiveRegionToUnionSetCoefficientMap>
+
+/**
+ * A list of sets of primitive regions, where the index of the list represents the reporting set ID,
+ * and the element of the list is the set of primitive regions which forms the corresponding
+ * reporting set. For example, if reportingSetId = 0, then the set contains the primitive regions
+ * that have "0"-bit position on, i.e. allPrimitiveRegionSetsList\[reportingSetId\] =
+ * setOf(1(=b’001’), 3(=b’011’), 5(=b’101’), 7(=b’111’)).
+ */
+private typealias VennDiagramDecomposition = List<Set<PrimitiveRegion>>
+
+/**
+ * A memory cache that stores the decompositions of Venn Diagrams given different number of
+ * reporting sets.
+ */
+private typealias VennDiagramDecompositionCache =
+  MutableMap<NumberReportingSets, VennDiagramDecomposition>
```

Comment:
I remember we did not have this memory in the alpha version. Is it a new feature we add in w2alpha?